File: DSP Simulation/goertzel_FFT.m
function pow = goertzel_FFT(data, k, N)
    omega = 2 * pi * k/N;
    cr = cos(omega);
    ci = sin(omega);
    coeff = 2 * cr;
    sprev = 0;
    sprev2 = 0;
    
    for i = 1:N
       s = data(i) + coeff * sprev - sprev2;
       sprev2 = sprev;
       sprev = s;
    end
    
    It = cr * sprev - sprev2;
    Qt = ci * sprev;
    
    omega2 = 2 * pi * k;
    cr2 = cos(omega2);
    ci2 = sin(omega2);
    
    I = It*cr2 + Qt*ci2;
    Q = -It*ci2 + Qt*cr2;
     
    pow = sqrt(I * I + Q * Q) * (2 / N);
    %pow = sprev2 * sprev2 + sprev * sprev - coeff * sprev * sprev2;
end

File: DSP Simulation/goertzel_FFT.cpp
#define _USE_MATH_DEFINES
#include <cmath>
#include <iostream>
using namespace std;

double goertzel_FFT(double* data, double k, int N) {
	
	double omega = 2.0 * M_PI * k / (double) N;
	double cr = cos(omega * M_PI / 180.0);
	double ci = sin(omega * M_PI / 180.0);
	double coeff = 2.0 * cr;
	double sprev = 0.0;
	double sprev2 = 0.0;

	for (int i = 0; i < N - 1; i++) {
		double s = data[i] + coeff * sprev - sprev2;
		sprev2 = sprev;
		sprev = s;
	}

	double It = cr * sprev - sprev2;
	double Qt = ci * sprev;

	double omega2 = 2.0 * M_PI * k;
	double cr2 = cos(omega2 * M_PI / 180.0);
	double ci2 = sin(omega2 * M_PI / 180.0);

	double I = It * cr2 + Qt * ci2;
	double Q = -It * ci2 + Qt * cr2;

	return sqrt(I * I + Q * Q);
}